control-plane-operator/hostedclusterconfigoperator/controllers/machine/machine.go

```diff
@@ -0,0 +1,137 @@
+	machine := &capiv1.Machine{}
+	if err := r.client.Get(ctx, client.ObjectKey(req.NamespacedName), machine); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "Machine", r)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get Machine: %w", err)
+	}
+
+	hcp := &hyperv1.HostedControlPlane{}
+	if err := r.client.Get(ctx, r.hcpKey, hcp); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found", "HostedControlPlane", r.hcpKey)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get HostedControlPlane: %w", err)
+	}
+	switch hcp.Spec.Platform.Type {
+	case hyperv1.KubevirtPlatform:
+		if hcp.Spec.Platform.Kubevirt != nil && (hcp.Spec.Platform.Kubevirt.BaseDomainPassthrough != nil && *hcp.Spec.Platform.Kubevirt.BaseDomainPassthrough) {
+			if err := r.reconcileKubevirtDefaultIngressEndpoints(ctx, hcp, machine); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed reconciling kubevirt default ingress passthrough service endpoints: %w", err)
+			}
+		}
+	}
+	log.Info("Reconciled Machine")
+	return reconcile.Result{}, nil
+}
+
+func (r *reconciler) reconcileKubevirtDefaultIngressEndpoints(ctx context.Context, hcp *hyperv1.HostedControlPlane, machine *capiv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+	var namespace string
+	if hcp.Spec.Platform.Kubevirt.Credentials != nil {
+		namespace = hcp.Spec.Platform.Kubevirt.Credentials.InfraNamespace
+	} else {
+		namespace = hcp.Namespace
+	}
+
+	// Manifests for infra/mgmt cluster passthrough service
+	cpService := hcpmanifests.IngressDefaultIngressPassthroughService(namespace)
+
+	cpService.Name = fmt.Sprintf("%s-%s",
+		hcpmanifests.IngressDefaultIngressPassthroughServiceName,
+		hcp.Spec.Platform.Kubevirt.GenerateID)
+
+	if err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(cpService), cpService); err != nil {
+		return fmt.Errorf("failed to get default ingress passthrow Service: %w", err)
+	}
+
+	// If there is a selector endpoints should not be generated
+	if len(cpService.Spec.Selector) > 0 {
+		return nil
+	}
+
+	if len(cpService.Spec.Ports) == 0 {
+		return fmt.Errorf("missing default ingress passthrow Service %s/%s ports", cpService.Namespace, cpService.Name)
+	}
+
+	ipv4EndpointSlice := &discoveryv1.EndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cpService.Namespace,
+			Name:      cpService.Name + "-" + machine.Name + "-ipv4",
+			Labels: map[string]string{
+				"kubernetes.io/service-name": cpService.Name,
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: machine.APIVersion,
+				Kind:       machine.Kind,
+				Name:       machine.Name,
+				UID:        machine.UID,
+			}},
+		},
+		AddressType: discoveryv1.AddressTypeIPv4,
+	}
+
+	result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, ipv4EndpointSlice, func() error {
+		err := r.kubevirtInfraClient.Get(ctx, client.ObjectKeyFromObject(ipv4EndpointSlice), ipv4EndpointSlice)
```

Comment:
You shouldn't have to get the endpoint slice here. The `CreateOrUpdate` function performs the Get and populates the object in the function.

I think this means you're code should look more like this.

```
ownerRef := config.OwnerRefFrom(cpService)
ipv4EndpointSlice := &discoveryv1.EndpointSlice{
		ObjectMeta: metav1.ObjectMeta{
			Namespace: cpService.Namespace,
			Name:      cpService.Name + "-" + machine.Name + "-ipv4",
}

result, err := r.CreateOrUpdate(ctx, r.kubevirtInfraClient, ipv4EndpointSlice, func() error {
        ownerRef.ApplyTo(ipv4EndpointSlize)
        if ipv4EndpointSlice.Labels = nil {
             ipv4EndpointSlice.Labels = map[string]string{}
        }
        ipv4EndpointSlice.Labels["kubernetes.io/service-name"] = cpService.Name
        ipv4EndpointSlice.AddressType: discoveryv1.AddressTypeIPv4

< remaining endpoint slice logic>
}

```

Reply:
I am creating one endpointslice per machine and ip family, so it's easier to map it to a Reconcile to a specific machine and also a scale down will clean them up since I am using Machine as the owner instead of Service (using "service-name" label is enough).